### PR TITLE
Add terms_set query

### LINKED ISSUE #26915: How to expose the new Lucene CoveringQuery in the query dsl?
Lucene 7.1 comes with a new query, called the [CoveringQuery](https://issues.apache.org/jira/browse/LUCENE-7621). This is neat query that is like a bool query with should clauses but instead of global minimum should match for all documents it allows a minimum should match per document. I wonder how this new query can best be exposed in the ES query dsl?

It can be exposed in the `bool` query itself (via a `minimum_should_match` field or something like that), but that can be confusing as `CoveringQuery` Lucene query doesn't understand must, filter or must_not clauses and then we run into difficult to explain search results. At this moment I lean towards exposing this a separate query in the query dsl. I'm just not sure what the best name should be for this query, but I think the structure should look something like this:

```
{
   "covering" : {
     "minimum_should_match_field": "my_field",
     "queries": [
        {
          # query1
        },
        {
          # query2
        },
       # etc
     ]
   }
}
```

The field specified in the `minimum_should_match_field` must be a number field with doc values not turned off. There is no need to worry about negative values, because the `CoveringQuery` treats a minimum should match of 0 and lower as 1.

I think that the `match` query should also get a new `type` named `covering` (or another name) and a `minimum_should_match_field` option that then under the hood uses the `CoveringQuery` instead of `BooleanQuery` or `PhraseQuery`.

### PR DESCRIPTION
**Pull request by martijnvg**
This query returns documents that match with at least one ore more
of the provided terms. The number of terms that must match varies
per document and is either controlled by a minimum should match
field or computed per document in a minimum should match script.

PR for #26915

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+    public List<?> getValues() {
+        return values;
+    }
+
+    public String getMinimumShouldMatchField() {
+        return minimumShouldMatchField;
+    }
+
+    public TermsSetQueryBuilder setMinimumShouldMatchField(String minimumShouldMatchField) {
+        this.minimumShouldMatchField = minimumShouldMatchField;
```
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java]: Should we throw an exception if a script is already provided ?

File: core/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+        return this;
+    }
+
+    public Script getMinimumShouldMatchScript() {
+        return minimumShouldMatchScript;
+    }
+
+    public TermsSetQueryBuilder setMinimumShouldMatchScript(Script minimumShouldMatchScript) {
+        this.minimumShouldMatchScript = minimumShouldMatchScript;
```
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java]: Same here ?

**Review by jimczi — COMMENTED** [review]
I left two minor comments, LGTM otherwise

File: core/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java (hunk 1 above)
Comment (martijnvg) [reply core/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java]: Thx for checking this out! Good point, I'll add extra validation here.

**Pull request merged**